pkg/ruler/remotequerier_test.go

```diff
@@ -117,6 +118,114 @@ func TestRemoteQuerier_QueryReq(t *testing.T) {
 
 }
 
+func TestRemoteQuerier_SendRequest(t *testing.T) {
+	const errMsg = "this is an error"
+	var (
+		successfulResponse = &httpgrpc.HTTPResponse{
+			Code: http.StatusOK,
+			Headers: []*httpgrpc.Header{
+				{Key: "Content-Type", Values: []string{"application/json"}},
+			},
+			Body: []byte(`{
+						"status": "success","data": {"resultType":"vector","result":[]}
+					}`),
+		}
+		erroneousResponse = &httpgrpc.HTTPResponse{
+			Code: http.StatusBadRequest,
+			Headers: []*httpgrpc.Header{
+				{Key: "Content-Type", Values: []string{"application/json"}},
+			},
+			Body: []byte("this is an error"),
+		}
+	)
+
+	tests := map[string]struct {
+		response        *httpgrpc.HTTPResponse
+		err             error
+		expectedError   error
+		expectedRetries bool
+	}{
+		"errors with code 5xx are retried": {
+			err:             httpgrpc.Errorf(http.StatusInternalServerError, errMsg),
+			expectedError:   httpgrpc.Errorf(http.StatusInternalServerError, errMsg),
+			expectedRetries: true,
+		},
+		"context.Canceled error is retried": {
+			err:             context.Canceled,
+			expectedError:   context.Canceled,
+			expectedRetries: true,
+		},
+		"gRPC context.Canceled error is retried": {
```

Comment:
I don't think we should retry in case of context canceled. In practice it will not be retried by the function because the backoff checks the context, but in this test it's retried due to how it's mocked (we don't cancel the context passed to the `sendRequest()` but we just return the context canceled).

Not sure how to proceed, but I think these 2 test cases are misleading. An option is the remove these 2 test cases, another option is to test the context cancellation case properly, which means cancelling the context passed to the `sendRequest()`.

Reply:
I have opted for canceling the context passed to `sendRequest()` and showing that the request is not retried.